crates/chain/src/keychain/txout_index.rs

```diff
@@ -559,32 +555,107 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+/// // case between 0 and a constant representing the maximum BIP32 derivation index.
+/// const BIP32_MAX_INDEX: u32 = (1 << 31) - 1;
+/// let _spk_range_iter = SpkIterator::new_with_range(&descriptor, 0..BIP32_MAX_INDEX);
+/// ```
+#[derive(Clone)]
+pub struct SpkIterator<D> {
+    next_index: u32,
+    end: u32,
+    descriptor: D,
+    secp: Secp256k1<bitcoin::secp256k1::VerifyOnly>,
+}
+
+impl<D> SpkIterator<D>
 where
-    R: Iterator<Item = u32> + Clone + Send + 'a,
+    D: Borrow<Descriptor<DescriptorPublicKey>>,
 {
-    let secp = Secp256k1::verification_only();
-    let has_wildcard = descriptor.has_wildcard();
-    range
-        .into_iter()
-        // non-wildcard descriptors can only have one derivation index (0)
-        .take_while(move |&index| has_wildcard || index == 0)
-        // we can only iterate over non-hardened indices
-        .take_while(|&index| index <= BIP32_MAX_INDEX)
-        .map(
-            move |index| -> Result<_, miniscript::descriptor::ConversionError> {
-                Ok((
-                    index,
-                    descriptor
-                        .at_derivation_index(index)
-                        .derived_descriptor(&secp)?
-                        .script_pubkey(),
-                ))
+    /// Creates a new script pubkey iterator starting at 0 from a descriptor.
+    pub fn new(descriptor: D) -> Self {
+        let end = if descriptor.borrow().has_wildcard() {
+            BIP32_MAX_INDEX
+        } else {
+            0
+        };
+
+        SpkIterator::new_with_range(descriptor, 0..=end)
+    }
+
+    /// Creates a new script pubkey iterator from a descriptor with a given range.
+    pub fn new_with_range<R>(descriptor: D, range: R) -> Self
+    where
+        R: RangeBounds<u32>,
+    {
+        Self {
+            next_index: match range.start_bound() {
+                Bound::Included(start) => *start,
+                Bound::Excluded(start) => *start + 1,
+                Bound::Unbounded => u32::MIN,
+            },
+            end: match range.end_bound() {
+                Bound::Included(end) => *end + 1,
+                Bound::Excluded(end) => *end,
+                Bound::Unbounded => u32::MAX,
             },
-        )
-        .take_while(Result::is_ok)
-        .map(Result::unwrap)
+            descriptor,
+            secp: Secp256k1::verification_only(),
+        }
+    }
+}
+
+impl<D> Iterator for SpkIterator<D>
+where
+    D: Borrow<Descriptor<DescriptorPublicKey>>,
+{
+    type Item = (u32, Script);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // For non-wildcard descriptors, we expect the first element to be Some((0, spk)), then None after.
+        // For wildcard descriptors, we expect it to keep iterating until exhausted.
+        let has_wildcard = self.descriptor.borrow().has_wildcard();
+        if self.next_index >= self.end || (has_wildcard && self.next_index > BIP32_MAX_INDEX) {
```

Comment:
You only need to check `self.next_index >= self.end`. `end` must be at most `BIP32_MAX_INDEX`.